File: src/main/java/pr/eleks/we_at_her/services/view/impl/ViewServiceImpl.java
package pr.eleks.we_at_her.services.view.impl;

import org.springframework.core.env.Environment;
import org.springframework.stereotype.Service;
import org.springframework.web.client.RestTemplate;
import org.springframework.web.util.UriComponentsBuilder;
import pr.eleks.we_at_her.dto.BlogPostDto;
import pr.eleks.we_at_her.dto.CityDto;
import pr.eleks.we_at_her.dto.UserDto;
import pr.eleks.we_at_her.dto.weather.WeatherSampleDto;
import pr.eleks.we_at_her.exceptions.PropertyNotFoundException;
import pr.eleks.we_at_her.services.EmailService;
import pr.eleks.we_at_her.services.view.ViewService;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Optional;

@Service
public class ViewServiceImpl implements ViewService {

    private Environment env;
    private RestTemplate restTemplate;
    private EmailService emailService;

    public ViewServiceImpl(Environment env, RestTemplate restTemplate, EmailService emailService) {
        this.env = env;
        this.restTemplate = restTemplate;
        this.emailService = emailService;
    }

    @Override
    public UriComponentsBuilder hostUriBuilder() throws PropertyNotFoundException {
        return UriComponentsBuilder
                .fromUriString(
                        Optional
                                .ofNullable(env.getProperty("server.host"))
                                .orElseThrow(() -> new PropertyNotFoundException("server.host"))
                                .concat(":").concat(
                                Optional
                                        .ofNullable(env.getProperty("server.port"))
                                        .orElseThrow(() -> new PropertyNotFoundException("server.port"))
                        )
                );
    }

    @Override
    public List<WeatherSampleDto> getAllWeatherSamples() throws PropertyNotFoundException {
        String requestUri = hostUriBuilder()
                .pathSegment("REST")
                .pathSegment("weatherSample")
                .toUriString();

        // Make request
        return new ArrayList<>(Arrays.asList(
                Optional
                        .ofNullable(restTemplate.getForObject(requestUri, WeatherSampleDto[].class))
                        .orElse(new WeatherSampleDto[0])
        ));
    }

    @Override
    public List<CityDto> getAllCities() throws PropertyNotFoundException {
        String requestUri = hostUriBuilder()
                .pathSegment("REST")
                .pathSegment("city")
                .toUriString();

        // Make request
        CityDto[] cityDtos = restTemplate.getForObject(requestUri, CityDto[].class);
        return new ArrayList<>(Arrays.asList(
                Optional
                        .ofNullable(cityDtos)
                        .orElse(new CityDto[0])
        ));
    }

    @Override
    public UserDto createUser(UserDto userDto) throws PropertyNotFoundException {
        String requestUri = hostUriBuilder()
                .pathSegment("REST")
                .pathSegment("user")
                .toUriString();

        UserDto savedUser = restTemplate.postForEntity(requestUri, userDto, UserDto.class).getBody();

        if (savedUser != null) {
            String activateUri = hostUriBuilder()
                    .pathSegment("activate")
                    .queryParam("u", savedUser.getUuid().toString())
                    .toUriString();

            emailService.sendEmail(savedUser.getEmail(),
                    "Your account was registered",
                    "Dear " + savedUser.getUsername() + "." +
                            "\nYour account was successfully created." +
                            "\nTo activate account please go to " + activateUri + " .");
        }
        return savedUser;
    }

    @Override
    public UserDto activateUser(String userUuid) throws PropertyNotFoundException {
        String requestUri = hostUriBuilder()
                .pathSegment("REST")
                .pathSegment("activateUser")
                .pathSegment(userUuid)
                .toUriString();

        return Optional
                .ofNullable(restTemplate.getForObject(requestUri, UserDto.class))
                .orElse(null);
    }

    @Override
    public List<BlogPostDto> getAllBlogPosts() throws PropertyNotFoundException {
        String requestUri = hostUriBuilder()
                .pathSegment("REST")
                .pathSegment("user")
                .toUriString();

        // Make request
        BlogPostDto[] blogPostDto = restTemplate.getForObject(requestUri, BlogPostDto[].class);
        return new ArrayList<>(Arrays.asList(
                Optional
                        .ofNullable(blogPostDto)
                        .orElse(new BlogPostDto[0])
        ));
    }
}

File: src/main/kotlin/pr/eleks/we_at_her/services/view/impl/ViewServiceImpl.kt
package pr.eleks.we_at_her.services.view.impl

import org.springframework.core.env.Environment
import org.springframework.stereotype.Service
import org.springframework.web.client.RestTemplate
import org.springframework.web.util.UriComponentsBuilder
import pr.eleks.we_at_her.dto.BlogPostDto
import pr.eleks.we_at_her.dto.CityDto
import pr.eleks.we_at_her.dto.UserDto
import pr.eleks.we_at_her.dto.weather.WeatherSampleDto
import pr.eleks.we_at_her.exceptions.PropertyNotFoundException
import pr.eleks.we_at_her.services.EmailService
import pr.eleks.we_at_her.services.view.ViewService

@Service
class ViewServiceImpl(
        private val env: Environment,
        private val restTemplate: RestTemplate,
        private val emailService: EmailService
) : ViewService {
    override fun hostUriBuilder(): UriComponentsBuilder =
            UriComponentsBuilder
                    .fromUriString((env.getProperty("server.host")
                            ?: throw PropertyNotFoundException("server.host"))
                            + ":" +
                            (env.getProperty("server.port")
                                    ?: throw PropertyNotFoundException("server.port"))
                    )

    override fun getAllWeatherSamples(): List<WeatherSampleDto?> {
        val requestUri = hostUriBuilder()
                .pathSegment("REST")
                .pathSegment("weatherSample")
                .toUriString()
        // Make request
        return restTemplate.getForObject(requestUri, Array<WeatherSampleDto?>::class.java)?.toList()
                ?: emptyList()
    }

    override fun getAllCities(): List<CityDto?> {
        val requestUri = hostUriBuilder()
                .pathSegment("REST")
                .pathSegment("city")
                .toUriString()
        // Make request
        return restTemplate.getForObject(requestUri, Array<CityDto?>::class.java)?.toList()
                ?: emptyList()
    }

    override fun createUser(userDto: UserDto): UserDto? {
        val requestUri = hostUriBuilder()
                .pathSegment("REST")
                .pathSegment("user")
                .toUriString()

        val savedUser = restTemplate.postForEntity(requestUri, userDto, UserDto::class.java).body
        if (savedUser != null) {
            val activateUri = hostUriBuilder()
                    .pathSegment("activate")
                    .queryParam("u", savedUser.uuid.toString())
                    .toUriString()
            emailService.sendEmail(savedUser.email!!,
                    "Your account was registered",
                    """
                    |Dear ${savedUser.username}.
                    |Your account was successfully created.
                    |To activate account please go to $activateUri .
                    """.trimMargin())
        }
        return savedUser
    }

    override fun getAllBlogPosts(): List<BlogPostDto?> {
        val requestUri = hostUriBuilder()
                .pathSegment("REST")
                .pathSegment("blogPost")
                .toUriString()
        // Make request
        return restTemplate.getForObject(requestUri, Array<BlogPostDto?>::class.java)?.toList()
                ?: emptyList()
    }

    override fun activateUser(userUuid: String): UserDto? {
        val requestUri = hostUriBuilder()
                .pathSegment("REST")
                .pathSegment("activateUser")
                .pathSegment(userUuid)
                .toUriString()

        return restTemplate.getForObject(requestUri, UserDto::class.java)
    }

    override fun createNewBlogPost(blogPostDto: BlogPostDto): BlogPostDto? {
        val requestUri = hostUriBuilder()
                .pathSegment("REST")
                .pathSegment("blogPost")
                .toUriString()

        return restTemplate.postForEntity(requestUri, blogPostDto, BlogPostDto::class.java).body
    }
}